samcli/lib/sync/sync_flow_factory.py

```diff
@@ -170,14 +175,45 @@ def _create_lambda_flow(
             )
         return None
 
-    def _create_layer_flow(self, resource_identifier: ResourceIdentifier, resource: Dict[str, Any]) -> SyncFlow:
-        return LayerSyncFlow(
-            str(resource_identifier),
-            self._build_context,
-            self._deploy_context,
-            self._physical_id_mapping,
-            self._stacks,
-        )
+    def _create_layer_flow(
+        self, resource_identifier: ResourceIdentifier, resource: Dict[str, Any]
+    ) -> Optional[SyncFlow]:
+        layer = self._build_context.layer_provider.get(str(resource_identifier))
+        if not layer:
+            LOG.warning("Can't find layer resource with '%s' logical id", str(resource_identifier))
+            return None
+
+        if BuildContext.is_layer_buildable(layer):
+            return LayerSyncFlow(
+                str(resource_identifier),
+                self._build_context,
+                self._deploy_context,
+                self._physical_id_mapping,
+                self._stacks,
+            )
+
+        if is_local_folder(layer.codeuri):
+            LOG.debug("Creating LayerSyncFlowSkipBuildDirectory for '%s' resource", resource_identifier)
+            return LayerSyncFlowSkipBuildDirectory(
+                str(resource_identifier),
+                self._build_context,
+                self._deploy_context,
+                self._physical_id_mapping,
+                self._stacks,
+            )
+
+        if is_zip_file(layer.codeuri):
+            LOG.debug("Creating LayerSyncFlowSkipBuildZipFile for '%s' resource", resource_identifier)
+            return LayerSyncFlowSkipBuildZipFile(
```

Comment:
Nit: This is how I read the name, which led to some confusion on my end. LayerSyncFlow + SkipBuild + Zip + File. Reading it this way, made me think we were zipping the file/folder but doesn't seem like that is the ended thing. Maybe we should rename or add in some "_" to signal to the reader the right thing from the name?

Reply:
Naming is hard and I am also not very happy with these names, open for any suggestion.